src/map/treasure_pool_container.cpp

```diff
@@ -0,0 +1,225 @@
+            }
+            return PTreasurePool;
+        }
+
+        // 4b. First party member in zone
+        auto& newTp = CreateTreasurePool(PPartyLeader->id, TREASUREPOOL_PARTY, TREASUREPOOL_MANAGED);
+        newTp.AddMember(PChar);
+        return newTp;
+    }
+
+    // 5. Find any solo pool we may have already created
+    if (m_Pools.contains(PChar->id))
+    {
+        return m_Pools.at(PChar->id);
+    }
+
+    // Otherwise, create a new solo pool
+    auto& newTp = CreateTreasurePool(PChar->id, TREASUREPOOL_SOLO, TREASUREPOOL_MANAGED);
+    newTp.AddMember(PChar);
+    return newTp;
+}
+
+auto CTreasurePoolContainer::GetTreasurePools() -> std::unordered_map<uint32, CTreasurePool>&
+{
+    return m_Pools;
+}
+
+// Reassigns the map key for a given treasure pool.
+// Used when party or alliance leader changes.
+auto CTreasurePoolContainer::ReassignTreasurePool(const CCharEntity* PPrev, const CCharEntity* PNew) -> bool
+{
+    if (!PPrev || !PNew)
+    {
+        ShowError("CTreasurePoolContainer::ReassignTreasurePool() - PPrev or PNew was null.");
+        return false;
+    }
+
+    if (auto node = m_Pools.extract(PPrev->id); !node.empty())
+    {
+        node.key() = PNew->id;
+        m_Pools.insert(std::move(node));
+        return true;
+    }
+
+    return false;
+}
+
+// Force releases a treasure pool from the container.
+// Used when dissolving a party or alliance.
+auto CTreasurePoolContainer::ReleaseTreasurePool(CTreasurePool& PTreasurePool) -> void
+{
+    for (auto it = m_Pools.begin(); it != m_Pools.end();)
+    {
+        if (&it->second == &PTreasurePool)
+        {
+            ShowDebugFmt("Releasing treasure pool owned by ID {}", it->first);
+            for (auto member : PTreasurePool.GetMembers())
+            {
+                PTreasurePool.DelMember(member);
+            }
+            it = m_Pools.erase(it);
+        }
+        else
+        {
+            ++it;
+        }
+    }
+}
+
+auto CTreasurePoolContainer::CreateTreasurePool(const uint32 poolOwnerId, const TREASUREPOOLTYPE PoolType, const TREASUREPOOLMANAGEMENT Management) -> CTreasurePool&
+{
+    ShowDebugFmt("Creating treasure pool owned by {} with initial capacity {}", poolOwnerId, static_cast<int>(PoolType));
+
+    if (PoolType == TREASUREPOOL_SHARED)
+    {
+        m_ZonePools.emplace(poolOwnerId, CTreasurePool(PoolType, Management));
+        return m_ZonePools.at(poolOwnerId);
+    }
+
+    m_Pools.emplace(poolOwnerId, CTreasurePool(PoolType, Management));
+    return m_Pools.at(poolOwnerId);
+}
+
+// Check for expired items and clean up empty pools.
+auto CTreasurePoolContainer::OnZoneTick(const time_point tick) -> void
+{
+    for (auto it = m_Pools.begin(); it != m_Pools.end();)
+    {
+        it->second.CheckItems(tick);
+
+        if (it->second.MemberCount() < 1)
+        {
+            ShowDebugFmt("Removing empty pool formerly owned by ID {}", it->first);
+            it = m_Pools.erase(it);
```

Comment:
I tend not to like this pattern of `I am going to iterate with raw iterators and remove things as I go`, can you change this to do similar to what we do in `ZoneEntities`: Where we collect the things we want to remove into a temporary container as we do the first pass, and then remove all the container things. In this case, then you'd flush afterwards.